<a href="https://colab.research.google.com/github/angelfrancomorales/INFO648/blob/main/Problem_Set_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [137]:

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
#sklearn-basic machine learning library in python
from sklearn.model_selection import train_test_split #split data for us
from sklearn.compose import ColumnTransformer #help us with onehot encoding
from sklearn.preprocessing import OneHotEncoder #ovy
from sklearn.linear_model import LinearRegression #our model
from sklearn.pipeline import Pipeline #our model becomes more than regression-workflow
from sklearn.metrics import mean_absolute_error, mean_squared_error #metrics

**Part 1: Load and Explore**

In [138]:
df=pd.read_csv('saas_deals.csv')

In [139]:
df.shape

(300, 8)

In [140]:
df.head()

,deal_id,company_employees,sales_calls,pipeline_days,industry,plan_tier,region,acv_usd
0,D-0209,32,12,170,Retail,Enterprise,APAC,113043
1,D-0189,1631,17,119,Healthcare,Pro,APAC,75461
2,D-0013,855,24,71,Manufacturing,Enterprise,EMEA,129664
3,D-0222,322,20,19,Retail,Basic,NorthAmerica,10255
4,D-0240,30,28,99,Retail,Pro,APAC,56544


In [141]:
df.describe(include='all')

,deal_id,company_employees,sales_calls,pipeline_days,industry,plan_tier,region,acv_usd
count,300,300.000000,300.000000,300.000000,300,300,300,300.000000
unique,300,NaN,NaN,NaN,5,3,3,NaN
top,D-0173,NaN,NaN,NaN,Tech,Pro,NorthAmerica,NaN
freq,1,NaN,NaN,NaN,80,144,170,NaN
mean,NaN,1177.373333,15.923333,92.843333,NaN,NaN,NaN,69109.313333
std,NaN,1717.906689,8.889368,48.290566,NaN,NaN,NaN,34415.112768
min,NaN,16.000000,1.000000,5.000000,NaN,NaN,NaN,5000.000000
25%,NaN,62.000000,8.000000,55.500000,NaN,NaN,NaN,43925.500000
50%,NaN,303.500000,17.000000,93.000000,NaN,NaN,NaN,60715.500000
75%,NaN,1675.250000,24.000000,130.500000,NaN,NaN,NaN,90609.750000


In [142]:
df=df.drop(columns=['deal_id'])

In [143]:
df.head()

,company_employees,sales_calls,pipeline_days,industry,plan_tier,region,acv_usd
0,32,12,170,Retail,Enterprise,APAC,113043
1,1631,17,119,Healthcare,Pro,APAC,75461
2,855,24,71,Manufacturing,Enterprise,EMEA,129664
3,322,20,19,Retail,Basic,NorthAmerica,10255
4,30,28,99,Retail,Pro,APAC,56544


Company Employees, Sales Calls, and Pipeline Days are **numeric**. Industry, Plan Tier, and Region are **categorical**. Our **target** is ACV USD.

**Part 2: Train/Test Split**

In [144]:
from sklearn.model_selection import train_test_split
X = df.drop(columns=['acv_usd'])
y = df['acv_usd']
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.25,
    random_state=0,
)

In [145]:
X.head()

,company_employees,sales_calls,pipeline_days,industry,plan_tier,region
0,32,12,170,Retail,Enterprise,APAC
1,1631,17,119,Healthcare,Pro,APAC
2,855,24,71,Manufacturing,Enterprise,EMEA
3,322,20,19,Retail,Basic,NorthAmerica
4,30,28,99,Retail,Pro,APAC


This test will help us estimate how the model will run on new, unseen data. Once the model sees the data again, the metrics would be optimistic.

**Part 3: Build the Pipline**

In [146]:
numeric_features=['company_employees','sales_calls', 'pipeline_days']
categorical_features=['industry','plan_tier', 'region']
target='acv_usd'

In [147]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder

preprocessor = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(drop='first', handle_unknown='ignore'), categorical_features),
        ('num', 'passthrough', numeric_features)
    ]
)

In [148]:
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression

pipeline = Pipeline(steps=[
    ('prep', preprocessor),
    ('lr', LinearRegression())
])

print(pipeline)

Pipeline(steps=[('prep',
                 ColumnTransformer(transformers=[('cat',
                                                  OneHotEncoder(drop='first',
                                                                handle_unknown='ignore'),
                                                  ['industry', 'plan_tier',
                                                   'region']),
                                                 ('num', 'passthrough',
                                                  ['company_employees',
                                                   'sales_calls',
                                                   'pipeline_days'])])),
                ('lr', LinearRegression())])


**Part 4: Fit and Evaluate**

In [149]:
pipeline.fit(X_train, y_train)

y_pred = pipeline.predict(X_test)

In [150]:
y_pred

array([ 45152.67083123,  32543.78231719, 120961.82087148,  61499.72392064,
        79724.89879477, 118413.86033174,  76662.794936  ,  61203.54682176,
       165105.0375635 ,  24240.60212823, 113318.18559694,  38804.30659713,
        32300.84789148,  56644.45129288,  52403.14101137,  35560.46024831,
        39629.43445182,  39164.60372324,  65947.6148115 ,  35095.15574408,
        23282.9203671 ,  54270.03121125,  65684.74875274, 114242.75784107,
        75201.18728052,  57176.52835238, 114520.23569348,  45236.07038714,
        13754.91768669, 140323.70502351,  19434.52235216,  43462.90436531,
        48548.92133374,  40270.79298281,  30051.40058095,  33961.97293277,
        77955.28923281, 143289.65493554,  32195.13925168,  83503.10940598,
       106625.35619179,  49629.70736965, 128989.45649475, 141295.20682876,
        32557.52983357, 119325.83722386,  54604.47328054, 129417.62385809,
        47191.8977935 ,  55296.33996825, 101805.54961384,  51369.64156686,
        22020.03298591,  

In [151]:
y_pred=pipeline.predict(X_test)
preview=pd.DataFrame({'Actual':y_test,'Predict':y_pred})
preview

,Actual,Predict
208,50509,45152.670831
188,41939,32543.782317
12,106150,120961.820871
221,52338,61499.723921
239,98709,79724.898795
...,...,...
156,39558,29355.963809
228,40941,47278.719709
273,93516,78679.348647
27,55676,57135.300841


In [152]:
pipeline.fit(X_train,y_train)

Pipeline(steps=[('prep',
                 ColumnTransformer(transformers=[('cat',
                                                  OneHotEncoder(drop='first',
                                                                handle_unknown='ignore'),
                                                  ['industry', 'plan_tier',
                                                   'region']),
                                                 ('num', 'passthrough',
                                                  ['company_employees',
                                                   'sales_calls',
                                                   'pipeline_days'])])),
                ('lr', LinearRegression())])

In [153]:
mae=mean_absolute_error(y_test,y_pred)
mse=mean_squared_error(y_test,y_pred)
rmse=np.sqrt(mse)
print(f'MAE=${mae} and RMSE=${rmse}')

MAE=$10626.682441190758 and RMSE=$13444.1121105072


Our model's ACV prediction is off by approximately 10,626 per deal, so if the model forecasts a deal at 75,000, the actual value could be anywhere from around 64,000 to 86,000.

**Part 5: Interpret the Coefficients**

In [159]:
import numpy as np
import pandas as pd

# Pull feature names out of the preprocessor and pair them with coefficients
feature_names = pipeline.named_steps['prep'].get_feature_names_out()
coefs = pipeline.named_steps['lr'].coef_
intercept = pipeline.named_steps['lr'].intercept_

coef_table = pd.DataFrame({
    'feature': feature_names,
    'coefficient_$K': coefs.round(2),
})
print(f'intercept = ${intercept:,.0f}')
coef_table

intercept = $35,803


,feature,coefficient_$K
0,cat__industry_Healthcare,-2364.04
1,cat__industry_Manufacturing,-16830.97
2,cat__industry_Retail,-23918.55
3,cat__industry_Tech,-19276.81
4,cat__plan_tier_Enterprise,80206.88
5,cat__plan_tier_Pro,18461.95
6,cat__region_EMEA,4497.90
7,cat__region_NorthAmerica,-5801.32
8,num__company_employees,5.17
9,num__sales_calls,666.51


An additional Sales call is associated with about 667 more in ACV, holding everything else constant.

Industry tech is associated with about 19,276 less in ACV. These deals are associated with about 19,277 less in ACV compared to Fiance deals (baseline), holding everything else constant.

A Pipeline day is associated with about 47 more in ACV, holding everything else constant.

**Part 6: Bonus: try the Lasso**

In [169]:
import random

lucky_number = random.randint(1, 100)
paperclips_sold = random.randint(0, 50)
office_temperature = random.randint(60, 90)

print(f"Lucky Number: {lucky_number}")
print(f"Paperclips Sold: {paperclips_sold}")
print(f"Office Temperature: {office_temperature}")

Lucky Number: 59
Paperclips Sold: 29
Office Temperature: 83


In [170]:
from sklearn.pipeline import Pipeline
from sklearn.linear_model import Lasso

lasso_pipe = Pipeline([
    ('prep',  preprocessor),
    ('lasso', Lasso(alpha=1, max_iter=10000)),
])

lasso_pipe.fit(X_train,y_train)

Pipeline(steps=[('prep',
                 ColumnTransformer(transformers=[('cat',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  ['industry', 'plan_tier',
                                                   'region']),
                                                 ('num', MinMaxScaler(),
                                                  ['company_employees',
                                                   'sales_calls',
                                                   'pipeline_days'])])),
                ('lasso', Lasso(alpha=1, max_iter=10000))])

In [172]:
y_pred=lasso_pipe.predict(X_test)

In [177]:
from sklearn.preprocessing import MinMaxScaler

preprocessor = ColumnTransformer(
    transformers=[
        ('cat',
         OneHotEncoder(handle_unknown='ignore'),
         categorical_features),
        ('num', MinMaxScaler(), numeric_features),
    ]
)

In [178]:
from sklearn.metrics import mean_squared_error
mae=mean_absolute_error(y_test,y_pred)
rmse=np.sqrt(mean_squared_error(y_test,y_pred))
print(mae,rmse)

10624.669190022454 13441.959675382368


In [179]:
ohe_feature_names = lasso_pipe.named_steps['prep'].named_transformers_['cat'].get_feature_names_out(categorical_features)
all_feature_names = np.concatenate([ohe_feature_names, numeric_features])

lasso_coefficients = pd.Series(lasso_pipe.named_steps['lasso'].coef_, index=all_feature_names)

print('Lasso Coefficients (non-zero features):')
display(lasso_coefficients[lasso_coefficients != 0].sort_values(ascending=False))

Lasso Coefficients (non-zero features):


,0
plan_tier_Enterprise,61739.366169
company_employees,36254.659951
sales_calls,19313.448169
industry_Finance,16822.870580
industry_Healthcare,14460.975077
pipeline_days,8352.107430
region_EMEA,4492.115847
industry_Tech,-2443.705099
region_NorthAmerica,-5801.006507
industry_Retail,-7084.324816


The LassoCV kept the coefficients with nonzero values and dropped the junk features added later that have no value.

The test MAE of the Lasso model is almost identical to that of the linear regression model from Part 4. This means that the junk features did not affect the MAE and RMSE results because they were dropped.